### ISSUE TITLE
Injecting thumbnail information using "add()" is making it harder for the model to learn.

### ISSUE THREAD

**Issue by ramayer** [ramayer/Elucidated-Diffusion-Model-vibe-coded#5]
https://claude.ai/share/afb781f2-88d5-4a7f-9fd2-08a91dbd4f6c

## User

I do have a question about how we're injecting the Thumbnail information into the other path with +.    

If the unconditioned path was drawing great faces with no thumbnail, and suddenly the thumbnail pushes that vector in a different direction, the rest of the network needs to learn to draw a good face from both the pushed and nonperturbed vector.   On the otherhand if it were injected using mean() instead of add(), might the thumbnail path learn to create the same "pretty face here" vector that the unconditioned path had learned before -- and then mean() becomes a no-op that is more compatible than add() would have been?

## claude

Good question, and I think you've actually spotted a real weakness in additive injection — but I don't think the fix is mean() exactly. Let me separate what's right about your intuition from where a fixed 50/50 mean would hurt you.

What's right: if the thumbnail encoder ever learns to output something that points in roughly the same direction as h already does (which is plausible — a correctly-denoised face's 64×64-level features and a downsampled LR thumbnail's features should correlate, especially at low sigma where the backbone's own estimate is already good), pure addition doesn't give you a no-op. It gives you h + h ≈ 2h — same direction, doubled magnitude. GroupNorm downstream re-normalizes scale within a channel group, so this isn't catastrophic, but it doesn't fully undo it either: normalizing a group's joint statistics isn't the same as recovering "what h alone would have looked like," especially since the doubled vector's direction is a specific mix of both, not cleanly separable back out. So your instinct that addition forces downstream layers to reconcile "pushed" vs "unperturbed" regimes rather than getting a free pass when the two would agree anyway — that's a legitimate concern, not a nitpick.

Where a fixed mean would hurt you: (h + inject) / 2 bakes in a hard ceiling — the thumbnail can never contribute more than 50% of the result, no matter how much you want it to dominate. And you specifically do want it to dominate at high sigma, where x_t is nearly pure noise and the thumbnail is close to the only real signal available. A fixed mean caps exactly the case where you need the most override power.

The actual synthesis of your idea: a learned, gated interpolation rather than either fixed operation — output is a convex combination where the gate itself decides how much weight the thumbnail gets, per-channel and per-sigma, rather than being stuck at a constant:

```python
output = (1 - gate) * h + gate * inject_content
```

This gets you everything you wanted from mean() — when inject_content ≈ h, the blend is a genuine no-op regardless of gate's value, since interpolating between two equal things gives that same thing back, not a doubled one — while keeping the ability to swing gate → 1 and let the thumbnail fully override h when sigma is high and h is unreliable. It also directly fixes the magnitude-compounding case you flagged: as gate grows, h's own contribution is explicitly turned down, not left at full strength and summed against — so when the two roughly agree, you get a clean identity-like combination, and when they disagree, h doesn't get to fight the injection at full weight.

This is a genuinely different mechanism, not a variant of what you have — right now gate scales inject before adding it to a full-strength h; here gate decides the tradeoff between the two, which lives one level higher in CNNEncoder.forward rather than inside ThumbnailEncoder alone (since it now needs h, not just the thumbnail):

```python

# ThumbnailEncoder.forward returns (content, gate) instead of a single value
def forward(self, thumbnail, t_emb):
    content = self.proj(self.net(thumbnail))
    gate = torch.sigmoid(self.to_gate(t_emb))          # per-channel, per-t, in [0,1]
    return rearrange("b c -> b c 1 1", gate).expand_as(content), content

# in CNNEncoder.forward, at the injection level:
gate, content = thumbnail_encoder_output
h = (1 - gate) * h + gate * content
```

Zero-init still works, just via sigmoid's bias rather than a literal zero: initialize to_gate.bias to something like -10 (weight still zero) so gate ≈ sigmoid(-10) ≈ 4.5e-5 at start — near-enough to zero that output ≈ h, matching the identity-at-init property you already have, without the earlier deadlock issue (since gradient into content's path is no longer multiplied by an exact zero).

I'd treat this as a next-restart experiment, not something to swap into the run that's currently 14 hours in and showing real trending improvement — you'd be stacking another variable on top of a run that's already telling you something useful. Let it keep going and watch whether those two remaining bins (cond@3σ, uncond@30σ) keep trending down or plateau; that's a cleaner signal than switching mechanisms mid-flight would give you.

**Comment by ramayer** [comment]
## User

Cool.   And should ThumbnailEncoder's null method should return a gate that means "ignore me"?


## Claude

Yes, exactly — and this is actually a cleaner situation than before, worth spelling out why.

With the gated interpolation, output = (1 - gate) * h + gate * content. For "no thumbnail" to mean exactly output = h, you need gate = 0 — and critically, once gate = 0, content is multiplied by zero and literally doesn't matter what it is. That's different from your original additive scheme, where the null path needed a specific learned value (null_feat) that had to actually equal zero to be inert. Here, correctness doesn't depend on content being any particular value at all — only on gate being pinned at zero.

So null() should return gate=0, and content can just be zeros too for tidiness, both as frozen buffers, not learned parameters — for the same reason as before: if gate's null value were a learned parameter, gradient from unconditional batches could nudge it away from exactly zero to whatever locally helps unconditional loss, reopening the same drift problem you already fixed once:

```python
class ThumbnailEncoder(nn.Module):
    def __init__(self, in_channels, hidden_ch, out_ch, emb_dim, num_layers=3):
        super().__init__()
        layers = []
        cur_ch = in_channels
        for _ in range(num_layers):
            layers.append(nn.Conv2d(cur_ch, hidden_ch, 3, padding=1))
            layers.append(nn.GroupNorm(_largest_valid_group_count(hidden_ch, 8), hidden_ch))
            layers.append(nn.GELU())
            cur_ch = hidden_ch
        self.net = nn.Sequential(*layers)

        self.proj = nn.Conv2d(hidden_ch, out_ch, 1)
        nn.init.zeros_(self.proj.weight)
        nn.init.zeros_(self.proj.bias)

        self.to_gate = nn.Linear(emb_dim, out_ch)
        nn.init.zeros_(self.to_gate.weight)
        nn.init.constant_(self.to_gate.bias, -10.0)   # gate ≈ sigmoid(-10) ≈ 0 at init

        self.out_ch = out_ch

    def forward(self, thumbnail, t_emb):
        content = self.proj(self.net(thumbnail))
        gate = torch.sigmoid(self.to_gate(t_emb))
        gate = rearrange("b c -> b c 1 1", gate).expand_as(content)
        return gate, content

    def null(self, batch_size, height, width, device, dtype):
        gate = torch.zeros(batch_size, self.out_ch, height, width, device=device, dtype=dtype)
        content = torch.zeros_like(gate)   # value is irrelevant once gate=0, but keep it clean
        return gate, content
```

And the call site becomes symmetric regardless of which branch fired:

```
gate, content = (
    self.thumbnail_encoder(thumbnail, t_emb) if thumbnail is not None and not drop
    else self.thumbnail_encoder.null(B, res, res, device=x.device, dtype=x.dtype)
)
h = (1 - gate) * h + gate * content
```

One thing worth being deliberate about: gate=0 exactly (a hard buffer) means the unconditional path's output is bit-identical to "no thumbnail encoder exists at all" — genuinely as strong a guarantee as freezing null_injection gave you before, just now covering both operands instead of one.
